[RM Chap8] Multi-Cloud Interactions Model - Draft (#2629)
cntt-n/CNTT: doc/ref_model/chapters/chapter08.md

The Cloud Services and the Cloud Resources Brokers provide value-added services in addition to the fundamental capabilities like service and resource discovery.  These Brokers are critical for a multi-cloud environment to function and utilise cloud specific plugins to perform the necessary activities. These Brokers can, for example, provision and manage environments with resources and services for Machine Learning (ML) services, Augmented/Virtual Reality, or specific industries.

<a name="8.2.4"></a>
### 8.2.4 Multi-Cloud Interactions Model
<a name="8.2.4.1"></a>
### 8.2.4.1 Introduction

> To realise a federated cloud requires definition and agreement of set of APIs. These APIs should allow each of the parties to interact cooperatively and need to cover the management layer: business management and service operations interactions; as well as the data plane, customer and user, transactions and conversational interfaces.

Reviewer: typo
```suggestion
To realise a federated cloud requires the definition and agreement on a set of APIs. These APIs should allow each of the parties to interact cooperatively and need to cover the management layer: business management and service operations interactions; as well as the data plane, customer and user, transactions, and conversational interfaces.
```

Author: Thank you @pgoyal01 
updated

> To realise a federated cloud requires definition and agreement of set of APIs. These APIs should allow each of the parties to interact cooperatively and need to cover the management layer: business management and service operations interactions; as well as the data plane, customer and user, transactions and conversational interfaces.

Reviewer: Could a single plane of glass (a broker) utilise the underlying cloud's APIs but with centralised monitoring and alerting so as to be able to take global-level actions particularly in situations with failures, load, resource constraints, etc.

Author: @pgoyal01 

As per comment above on CaaS:
This needs refinement and thought, particularly around how Fault and Performance Management should be managed in these scenarios, which I have allowed for under: Manage Resource - Monitor Resources.

I would like your hear your thoughts on Broker and "single plane of glass".

In typical use "Broker" is external party acting as intermediary between the other parties.

This was the role that OASIS was initially aiming to support with its set of standards (including TOSCA), but the pattern has not really taken off in provision of Cloud Services, to a great extend. The main models being direct consumption and reseller.

In the top figure: "Figure 8-3 - Conceptual Architecture of a Telco Operator Platform" which had the "Cloud Service Broker" and the "Cloud Resource Broker" I always saw these as Orchestration functions, bringing together individual "sets of virtualised cloud resources" (or VIMs of both OpenStack and Kubernetes variety) and playing role of the Brokers outlined.

I think your question, could be answered with a yes, but this is a choice based on the extent to which the consumer wishes to control everything centrally vs delegate the control away.

Could I specify the desired desired behavior globally, but get the manage of this to be automatically distributed down to the various delegated layers ?

Point for discussion.

Author: @pgoyal01 @walterkozlowski  I have updated Multi-Cloud Interaction Model to include two different Brokerage models:
- API Brokerage
- Cloud Brokerage , which provide "Single Pane of Glass".

> As outlined in "Figure 8-3 - Conceptual Architecture of a Telco Operator Platform" above the exposure point for the Management Interactions is the "Cloud Service Broker" and the "Cloud Service Broker". The set of interactions that these interface points need to provide are defined by the "Multii-Cloud Interaction Model". This provides a taxonomy for the interactions and is illustrated below:

Reviewer: typo
```suggestion
As outlined in "Figure 8-3 - Conceptual Architecture of a Telco Operator Platform" above the exposure point for the Management Interactions is the "Cloud Service Broker" and the "Cloud Resources Broker". The set of interactions that these interface points need to provide are defined by the "Multii-Cloud Interaction Model". This provides a taxonomy for the interactions and is illustrated below:
```

Author: Thank you @pgoyal01 
Updated

> As outlined in "Figure 8-3 - Conceptual Architecture of a Telco Operator Platform" above the exposure point for the Management Interactions is the "Cloud Service Broker" and the "Cloud Service Broker". The set of interactions that these interface points need to provide are defined by the "Multii-Cloud Interaction Model". This provides a taxonomy for the interactions and is illustrated below:

Reviewer 2: ```suggestion
As outlined in "Figure 8-3 - Conceptual Architecture of a Telco Operator Platform" above the exposure point for the Management Interactions is the "Cloud Service Broker" and the "Cloud Service Broker". The set of interactions that these interface points need to provide are defined by the "Multi-Cloud Interaction Model". This provides a taxonomy for the interactions and is illustrated below:
```
Typo Multii

Reviewer: @ulikleber There is also another suggested change above - one of the duplicates "Cloud Service Broker" needs to be "Cloud Resources Broker".

Reviewer 2: @pgoyal01 I saw it, but didn't know how to do it properly. Trying again ;-) by adding your change manually.

```suggestion
As outlined in "Figure 8-3 - Conceptual Architecture of a Telco Operator Platform" above the exposure point for the Management Interactions is the "Cloud Service Broker" and the "Cloud Resources Broker". The set of interactions that these interface points need to provide are defined by the "Multi-Cloud Interaction Model". This provides a taxonomy for the interactions and is illustrated below:
```

Reviewer: @ulikleber Thanks

Author: @pgoyal01 & @ulikleber Thank you both, I believe I have corrected both errors.

The set of high level interactions cover:
- Manage Account - covering Account, Users, Billing
- Manage Connectivity - Public or Private Network, VPN Configuration, CSP Edge / Cloud Connection Configuration, Connection Security Profile
- Manage Resource - Resource Pool Management, VM / VNF Management (CPU, Memory, Storage, Network), Image Repository Management, Storage Management, VNF / CNF LCM, Monitor Resources
- Manage App/VNF - Image / Container / Registry Management, Deploy/Configure/Scale/Start/Stop App/VNF, Monitor App/VNFs
- Transactions / Conversations - Use Communications Services, Use Edge Applications Services, Use Cloud Services

<a name="8.2.4.2"></a>
### 8.2.4.2 Stereo-Typical Scenarios
A set of stereo-typical interactions cases are illustrated for the cases of a Simple Infrastructure-as-a-Service (IaaS) and Software-as-a-Service (SaaS), where deployment is on a Cloud Provider's centralised sites and/or Edge sites. The scenarios help highlight needs for the Cloud Service Broker and Cloud Resources Broker (as per Figure 8.3) and hence extent of orchestration required to manage the interactions.

> <p align="center"><img src="../figures/rm-chap8-multi-cloud-interactions-simple-stereo-types-02.png" alt="Simple Stereo-Type Interactions"></p>

Reviewer: Typo
```suggestion
<p align="center"><img src="../figures/rm-chap8-multi-cloud-interactions-simple-stereo-types-02.png" alt="Simple Stereo-Typical Interactions"></p>
```

Author: Thank you @pgoyal01 
Your english correction is correct.

> <p align="center"><b>Figure 8-5:</b> Simple Stereo-Type Interactions</p>

Reviewer: ```suggestion
<p align="center"><b>Figure 8-5:</b> Simple Stereo-Typical Interactions</p>
```

Author: Thank you @pgoyal01 

Updated.

The following patterns are visible:
- For IaaS Cloud Integration:

>   - Cloud behaves like VIM and so needs Orchestration

Reviewer: typo
```suggestion
  - Cloud behaves like a set of virtual servers and, thus, requires virtual server life-cycle management and orchestration
```

Author: Thank you @pgoyal01 , agree good to avoid acronyms, especially when undefined.

  - Depending on whether the cloud is accessed via public internet or private connection will change the extend of the Connectivity Management
- For SaaS Cloud Integration:

>   - Cloud behaviors is like subscription management, so avoids need for complex orchestration

Reviewer: ```suggestion
  - Cloud behaves like a running application/service and requires subscription management, and complex orchestration of the app/service and underlying resources
```

Author: Thank you @pgoyal01.
I have added to this sentence:

Cloud behaves like a running application/service and requires subscription management, and complex orchestration of the app/service and underlying resources is managed by SaaS provider with the User relieved of having to provide direct control of resources

As what I was saying is that at its most basis the consumption of SaaS takes away the users need to have to worry about the resource management (it is all handled behind the interface).

I know that this is a generalisation, as is most of the "stereo-typical" analysis but it is important on how the APIs would be presented to User, as level of control with SaaS is highly constrained.

Reviewer 3: @jbhartley, "level of control with SaaS is highly constrained": I couldn't agree more.  Especially from the operations and support perspective, the aspect of which often people forget, thinking of SaaS as a piece of software like MS Word.  I have some recent first hand experience ...

Author: @walterkozlowski acknowledged. Thank you Walter

- For CaaS Cloud Integration:

>   - Registry for pulling Containers could be from:

Reviewer: wouldn't containers be always closer to the IaaS model? At the SaaS level, the need is for resources which can be VM or containers?

Author: @pgoyal01

This was one of my observations where if the CaaS provided a restricted set of available containers from its own managed registry then behavior become more SaaS like (with constrained the Registry selection become equivalent to having a menu of Apps that could be available and hence very SaaS like).
For more general CaaS there you have the ability to push any Container up into the Registry then yes it then behaves more like IaaS....

This needs refinement and thought, particularly around how Fault and Performance Management should be managed in these scenarios, which I have allowed for under: Manage Resource - Monitor Resources.

Feedback and thoughts appreciated.

    - Cloud in which case consumption model is closer to SaaS or
    - from Private / Public Registry in which case integration model requires specific registry management elements
- For Edge Cloud Integration:
  - Adds need for Communications Service Provider and Cloud Provider physical, network underlay and overlay connectivity management

A disaggregated scenario for a CSP using SaaS who uses IaaS is illustrated in the following diagram:

> <p align="center"><img src="../figures/rm-chap8-multi-cloud-interactions-disaggregated-stereo-type-01.png" alt="Disaggregated SaaS Stereo-Type Interaction"></p>

Reviewer: In the diagram, why would there not be any connectivity management by the SaaS on the IaaS and to the IaaS?

```suggestion
<p align="center"><img src="../figures/rm-chap8-multi-cloud-interactions-disaggregated-stereo-type-01.png" alt="Disaggregated SaaS Stereo-Typical Interaction"></p>
```

Author: Thank you @pgoyal01 

Corrected.

Author: @pgoyal01 I have now updated the interactions diagrams to make the "Data / User Plane" network explicitly visible and this also highlights where underlying network exposed.

If the "SaaS" is "infrastructure-less" then they would be dependent on the underlying IaaS for the "Data / User Plane". I have updated diagram to make this explicit.

> <p align="center"><b>Figure 8-5:</b> "Disaggregated SaaS Stereo-Type Interaction</p>

Reviewer: ```suggestion
<p align="center"><b>Figure 8-5:</b> "Disaggregated SaaS Stereo-Typical Interaction</p>
```

Author: Thank you @pgoyal01 

Updated.

<a name="8.2.4"></a>
### 8.2.4 Multi-Cloud Interactions Model
<a name="8.2.4.1"></a>
### 8.2.4.1 Introduction

> To realise a federated cloud requires the definition and agreement on a set of APIs. These APIs should allow each of the parties to interact cooperatively and need to cover the management layer: business management and service operations interactions; as well as the data plane, customer and user, transactions and conversational interfaces.

Reviewer 4: These API-s are not yet defined or implemented, therefore Anuket specification projects can only refer these as gaps. 
If Anuket would like to define these API-s I would propose to create a new project for them in Anuket. For the implementation of these API-s in the different cloud solutions there will be a need to direct contributions to Kubernetes and OpenStack.

Author: @CsatariGergely  yes agree. The closing section will provide survey of what is available and the applicability of these to support the alternate scenarios.

The focus here is not to specify standards. it is to identify needs and highlight know gap, as per your feedback.

Thank you for feedback.

Reviewer: @jbhartley Can you add a Gap in Chapter 10 -- a new Section 10.3.6 API Gaps, and a reference to it as in, " ... only refer these as gaps (see [Section 10.3.6](./chapter10.md#10.3.6))."

@CsatariGergely would that address your concern?

Author: @pgoyal01 @CsatariGergely 

I have added Multi-Cloud Gap section into Chapter 10 to cover API Gaps. This is now part of this pull request.

Reviewer: @jbhartley Thanks

As outlined in "Figure 8-3 - Conceptual Architecture of a Telco Operator Platform" above the exposure point for the Management Interactions is the "Cloud Service Broker" and the "Cloud Resource Broker". The set of interactions that these interface points need to provide are defined by the "Multi-Cloud Interaction Model". This provides a taxonomy for the interactions and is illustrated below:

> <p align="center"><img src="../figures/rm-chap8-multi-cloud-interactions-02.png" alt="Multi-Cloud Interactions Model"></p>

Reviewer 4: Please share the source of the figure in the [figures folder](https://github.com/cntt-n/CNTT/tree/master/doc/ref_model/figures).

Author: @CsatariGergely diagrams are here: https://github.com/jbhartley/CNTT/blob/master/doc/ref_model/artefacts/anuket-ref-model-chap-8-multicloud-interactions-diagrams-01.pptx

<p align="center"><b>Figure 8-4:</b> Multi-Cloud Interactions Model</p>

The model defines the following core roles:
- Communications Service Provider (CSP) - is the party responsible for providing end user service to their customer
- Customer / User - are the parties that use the service (User) and establishes the business agreement for the service provision (Customer). For retail services the customer and user are the same party, while for enterprise services the Enterprise is the Customer (responsible for the business agreement) and its representatives are the Users.

> - Cloud Providers - are the parties providing the cloud services. These services could be any XaaS service. It could be that a CSP has an agreement with a SaaS Cloud, which in turn uses an IaaS Cloud Provider to deliver their service.

Reviewer 4: How match do we have SaaS in the scope of Anuket. I'm not aware of any production ready SaaS solution.

Author: @CsatariGergely @walterkozlowski

your question: "How much do we have SaaS in the scope of Anuket?" - I think this should be discussion for next meeting:

For Cloud Provider perspective it is important that SaaS Providers needs are supported.

For the "Multi-Cloud Interactions Model" as looking at general "XaaS" cases and in looking at "Brokerage" across Cloud Providers it is important that SaaS Cloud Provider consumption is considered to ensure Interactions have broad coverage and can be "paper verified" to cover broad range of scenarios.

On your point: "I'm not aware of any production ready SaaS solution" , could you please elaborate, as there is not question that there are lots of existing "production ready SaaS solution" I believe that you are making this point in a much more constrained and specific context.

Thank you for feedback.

Reviewer: @CsatariGergely Figure 8-3 covers SaaS, PaaS and IaaS. As you know in the hyperscalar world there is no pure IaaS. And as we start to create complex services, we may see them as consisting of NFs that are on IaaS/PaaS and a set of workloads (and tools) that are SaaS.

The set of high level interactions cover:
- Manage Account - covering Account, Users, Subscription, Billing
- Manage Connectivity - Public or Private Network, VPN Configuration, CSP Edge / Cloud Connection Configuration, Connection Security Profile
- Manage Resource - Resource Pool Management, VM / VNF Management (CPU, Memory, Storage, Network), Image Repository Management, Storage Management, VNF / CNF LCM, Monitor Resources
- Manage App/VNF - Image / Container / Registry Management, Deploy/Configure/Scale/Start/Stop App/VNF, Monitor App/VNFs
- Transactions / Conversations - Use Communications Services, Use Edge Applications Services, Use Cloud Services

<a name="8.2.4.2"></a>
### 8.2.4.2 Stereo-Typical Scenarios
A set of stereo-typical interactions cases are illustrated for the cases of a Simple Infrastructure-as-a-Service (IaaS) and Software-as-a-Service (SaaS), where deployment is on a Cloud Provider's centralised sites and/or Edge sites. The scenarios help highlight needs for the Cloud Service Broker and Cloud Resources Broker (as per Figure 8.3) and hence extent of orchestration required to manage the interactions.

> <p align="center"><img src="../figures/rm-chap8-multi-cloud-interactions-simple-stereo-types-02.png" alt="Simple Stereo-Typical Interactions"></p>

Reviewer 4: Please share the source of the figure in the [figures folder](https://github.com/cntt-n/CNTT/tree/master/doc/ref_model/figures).

The following patterns are visible:
- For IaaS Cloud Integration:
  - Cloud behaves like a set of virtual servers and, thus, requires virtual server life-cycle management and orchestration
  - Depending on whether the cloud is accessed via public internet or private connection will change the extend of the Connectivity Management
- For SaaS Cloud Integration:
  - Cloud behaves like a running application/service and requires subscription management, and complex orchestration of the app/service and underlying resources is managed by SaaS provider with the User is relieved of having to provide direct control of resources
- For CaaS Cloud Integration:
  - Registry for pulling Containers could be from:
    - Cloud in which case consumption model is closer to SaaS or
    - from Private / Public Registry in which case integration model requires specific registry management elements
- For Edge Cloud Integration:
  - Adds need for Communications Service Provider and Cloud Provider physical, network underlay and overlay connectivity management

A disaggregated scenario for a CSP using SaaS who uses IaaS is illustrated in the following diagram:

> <p align="center"><img src="../figures/rm-chap8-multi-cloud-interactions-disaggregated-stereo-type-01.png" alt="Disaggregated SaaS Stereo-Typical Interaction"></p>

Reviewer 4: Please share the source of the figure.

Author: @CsatariGergely diagrams are here: https://github.com/cntt-n/CNTT/blob/master/doc/ref_model/artefacts/anuket-ref-model-chap-8-multicloud-interactions-diagrams-01.pptx

- For CaaS Cloud Integration:
  - Registry for pulling Containers could be from:
    - Cloud in which case consumption model is closer to SaaS or
    - from Private / Public Registry in which case integration model requires specific registry management elements
- For Edge Cloud Integration:
  - Adds need for Communications Service Provider and Cloud Provider physical, network underlay and overlay connectivity management

A disaggregated scenario for a CSP using SaaS who uses IaaS is illustrated in the following diagram:

<p align="center"><img src="../figures/rm-chap8-multi-cloud-interactions-disaggregated-stereo-type-01.png" alt="Disaggregated SaaS Stereo-Typical Interaction"></p>
<p align="center"><b>Figure 8-5:</b> "Disaggregated SaaS Stereo-Typical Interaction"</p>

In disaggregated SaaS scenario the application provider is able to operate as an "infra-structureless" organisation. This could be achieved through SaaS organisation using public IaaS Cloud Providers which could include the CSP itself. A key consideration for CSP in both cloud provision and consumption in Multi-Cloud scenario is how to manage the integration across the Cloud Providers.

> To make this manageeable and avoid having to manage integration there are a number of models:

Reviewer 5: ```suggestion
To make this manageable and avoid integration, there are a number of models:
```

> - Industry Standard APIs that allow consistent consumption across Cloud Providers,

Reviewer 5: Which agreed Industry Standard APIs exist for that, or which SDO works on it and has chance to get adoption?

> - API Brokage which provide consistent set of Consumer facings APIs that manage adaption to prorietry APIs

Reviewer 5: What would be some appropriate way to mention [Cloud Management Platforms](https://en.wikipedia.org/wiki/Cloud_management#Cloud_Management_Platforms_(CMP)) here?
